README.md
# fftGen

Code generator to create code for a Fast Fourier Transform with the target to
speed up the transform by unrolling its internal loops together with additional
optimization measures.

Version 1 2021-Jun-21st jost


## Features

 * Generate C code to compute a Fast Fourier Transform (FFT) and an Inverse Fast
   Fourier Transform (IFFT).
 * Target of the generated code is to compute the algorithms in a very fast way.
   This is achieved by

   1. unrolling the loops,

   2. removing operations where sequence elements are multiplied by sine or
      cosine function values being zero or one,

   3. optionally removing operations which are obsolete in certain
      circumstances, see documentation for details.

   Note that optimization 1. comes at the expense of code size.


## <a id="Doc">Documentation</a>

A detailed user manual is provided as HTML file
[**doc/html/index.html**](doc/html/index.html). This file needs
`doc/html/doxy.css` for correct display.


## Code Size

Unrolling the loops increases the code size drastically. The total number of
code lines for a certain number `n` of FFT data points is

    n       Total LOC
    8       84
    16      228
    32      552
    64      1320
    128     3024
    256     6844
    512     15264
    1024    33696

Some of the optimizations reduce the number of code lines. For details on the
size of the generated source code see the [documentation](#Doc).


## <a id="Int">Integration</a>

The program generates the FFT code only, no function header and no function
ending. Also no definition or declaration of required variables is generated.
See the [documentation](#Doc) for details how to obtain according appropriate
code.

Advantage of this approach is that no assumption is required for the generated
code regarding the concrete type of the involved variables. The arrays and
variables can therefore be of any valid floating point type, e.g. `double` or
`float`.

The generated code must be copied manually into the source code of the target
program or included by an `#include` pre-processor statement. Details are
described in the documentation.


## <a id="Configuration">Configuration</a>

The number of FFT data points, the type of FFT (standard or inverse) and the
kind of optimization can be chosen using command line options when calling the
program, for details see the [documentation](#Doc).

But there are two additional options to configure the program. These have to be
configured when compiling the program on the compiler command line.

1.  The generated code is, as described in the [previous section](#Int),
    independent from the type of the involved variables. However, the generated
    literal constants do not have any type specifiers. They are thus of the
    default `double` type (usually 64-bit precision). According to the C
    standard are the arithmetic expressions in that case also to be realized in
    `double` arithmetic by the compiler.

    If the involved variables are defined as `float` (usually 32-bit precision)
    then to increase speed and save space some compilers might realize the
    arithmetic (as meant) also as `float` arithmetic because there is no other
    type involved beside the literal constants.

    However, the appearance of the literal constants can be changed to be on the
    save side in those cases. The format of those literal constants is defined
    internally in the source code `fftGen.c` by the pre-processor constant
    **NUMBER_FORMAT**. This constant has by default the value `"%21.14e"`,
    meaning double precision constants in scientific notation (`e`) with 14
    fractional digits and a total width of 21 characters.

    If the constants should e.g. be written as single precision ("`float`")
    constants with scientific notation, 7 fractional digits (usually sufficient
    for single precision) and a total width of 14 characters then the according
    format should look like `"%14.7ef"` (the trailing `f` denotes `float`). This
    will reduce the size of the source code as well as the size of the compiled
    program and increase the computation speed – compromising, however, the
    precision.

    To let this change be active the program must be recompiled with the new
    format specified on the compiler command line, usually using option `-D...`.

    For the described example the compiler command line should therefore
    additionally contain the argument `-DNUMBER_FORMAT='"%14.7ef"'`.

    If you don't use the GNU C compiler check your compiler's user manual on the
    exact way to specify pre-processor constants on the compiler command line.

2.  The generated code is by default not indented. However, indentation is
    sometimes desired, usually to increase readability. It is therefore possible
    to let `fftGen` indent the code.

    The indentation is internally in `fftGen`'s source code `fftGen.c` defined
    by pre-processor constant `INDENT`. This constant is by default empty, but
    it can also be changed on the compiler command line when `fftGen` is
    compiled.

    To let the generated code for example be indented by four spaces the
    compiler command line should additionally contain the argument
    `-DINDENT='"    "'` (in case of the GNU C compiler).

`fftGen` can be compiled using the provided [`Makefile`](Makefile) or the as
well provided shell script [`makefftgen.sh`](makefftgen.sh). To change the
configuration as described above edit that script and call it to recompile the
program.


## Deficiencies, Optional Improvements

It has been decided to precompute the numerous sine and cosine function values
required in the FFT and implement them as literal constants for speed.

But it would also be possible to store those precomputed constants in a
`static const` array. That could save a lot of space in the executable program
(but not necessarily – depending on the compiler).

However, the effect on the computation speed depends also on the compiler. The
indirect access could last longer than the immediate addressing, it could as
well also be equally fast (or even be a bit faster).

If the sine and cosine function values should be stored in a `const` array
instead being literal constants then the chosen design needs to be changed. It
is in that case no more possible to have only one type independent junk of
generated code as described in section [Integration](#Int).


## Building

The program is provided as single source code file [**`fftGen.c`**](fftGen.c).
A distribution of `fftGen` comes also with the file [**`fftGen`**](fftGen), a
compiled executable for an **x86_64** *Linux* system. Executables for most other
systems can be produced with a standard C compiler.

Building the executable and the documentation can most easily be done with the
[GNU Make Utility](https://www.gnu.org/software/make/). Open a command line
shell and change to the subdirectory `fftGen`. Type

    make help

to get an overview over the different available make targets.

### Compilation

You need a standard C99 C compiler to compile the program. The program has been
developed with the [GNU C compiler](https://gcc.gnu.org/).

The program is easiest compiled using the GNU Make Utility with:

    make fftGen

You can also call the provided shell script [`makefftgen.sh`](makefftgen.sh) to
compile the program. To adapt the format of the generated code edit that script
as described in section [Configuration](#Configuration).

### Creating the User Manual

The source of the user manual is also fftGen.c. To get the user manual
[Doxygen](http://www.stack.nl/~dimitri/doxygen/) and
[Perl](https://www.perl.org/) must be installed on your system.

To compile the documentation to an HTML user manual type

    make html

on the command line. The resulting user manual
[**doc/html/index.html**](doc/html/index.html) will be created in subdirectory
`doc/html`.

### Installation / Deinstallation

`fftGen` does not contain (and does not need) a particular installation
procedure. Just unpack the distribution archive to an arbitrary location on your
hard disk. Copy file `fftGen` (or the file created by your compiler if you
compiled `fftGen` new) to the location where you want to run it or to a
directory included in your `PATH` environment variable. This should also be
possible on Windows – you have to compile it for Windows though.

To deinstall `fftGen` just remove the files which have been extracted from the
distribution archive.


## <a id="Test">Test</a>

A regression test is provided. Start the test on the command line with

    make check

The test script needs the program [**test/fftTest.c**](test/fftTest.c). This
program runs the actual tests and will during the test be recompiled with
different options for every test case. Hint: `fftTest.c` might contain some
helpful constructions on the usage of `fftGen`.

If the tests finishes successfully it ends with message "test passed". Otherwise
an error message is displayed. The test results – `stdout.log` and `stderr.log`
– are created in subdirectory [**test**](test) as the messages logged from
`stdout` and `stderr`. References for both file can be found in
[**test/refs**](test/refs). The test script automatically compares the created
log files with those references after each test and reports any differences.

An additional result of the test is an HTML coverage report
[`test/fftGen.cov.html`](test/fftGen.cov.html). This file is a listing of the
source code of `fftGen`.

The test needs the GCC compiler with `gcov` to create the code instrumentation
and the coverage report. `Perl` is required to created the coverage report in
HTML and the GNU `diff` utility is required to compare the test result log files
with the references.


## License

GNU General Public License Version 3, see file [COPYING](COPYING) or
<https://www.gnu.org/licenses/>.


CHANGES

Version 1

- Initial release


test/fftTest.c
//##############################################################################
// File: fftTest.c
//
// Program to test the result of fftGen
//
// The program is very flexibly configurable using define options ("-D...") on
// the compiler command line.
//
// Version 1     2021 Jun 21st   jost.brachert@gmx.de
//
//------------------------------------------------------------------------------
// Copyright (C) 2021  Jost Brachert, jost.brachert@gmx.de
//
// This program is free software: you can redistribute it and/or modify it under
// the terms of the GNU General Public License as published by the Free Software
// Foundation; either version 3 of the license, or (at your option) any later
// version.
//
// This program is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
// FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
// details.
//
// You should have received a copy of the GNU General Public License along with
// this program, see file COPYING. If not, see https://www.gnu.org/licenses/.
//

#include  <stdio.h>
#include   <math.h>     // sin(),cos(),sqrt(),fabs(),atan()
#include <stdlib.h>     // rand(),RAND_MAX

#define  LOGO   "fftTest"

// Commented out defines come from compiler command line

//#define  M   5
#define  N  (1<<M)

//#define  REAL_IN_OPTIMIZED
//#define  REAL_OUT_OPTIMIZED
//#define  SYMM_IN_OPTIMIZED
//#define  SYMM_OUT_OPTIMIZED
//#define  TEST_OUTPUT
//#define  NON_ZERO_IMAG_INPUT
#ifndef FFT_TYPE
#define  FFT_TYPE       double
#endif

#ifndef EPS
#define  EPS   1.e-8                // Tolerance for comparison to reference
#endif

typedef  struct Complex {double  r;
                         double  i;}  COMPLEX;

FFT_TYPE  xr[N];
FFT_TYPE  xi[N];
COMPLEX  xRef[N];
COMPLEX  xOri[N];


void  fftRef (COMPLEX*, int);
void  fft (FFT_TYPE*,FFT_TYPE*);
void  ffti (FFT_TYPE*,FFT_TYPE*);
void  conv (COMPLEX, double*, double*);



int  main ()
{
    int  failed = 0;                // Flag: !=0: Test failed
    int  i;

    //==========================================================================
    // Create input vector

    double  ts = 1./N;              // Sampling time [s]

    for (i=0; i<N; ++i) {
               // Ampl                         Phase
        xr[i] =   0.1*cos ((2.*M_PI)*1.*i*ts + 3.0)     // f=1/N Hz
                + 0.2*cos ((2.*M_PI)*2.*i*ts + 2.0)     // f=2/N Hz
                + 0.3*cos ((2.*M_PI)*3.*i*ts + 1.0)     // f=3/N Hz
#if M > 3
                + 0.4*cos ((2.*M_PI)*4.*i*ts + 0.0)     // f=4/N Hz
                + 0.5*cos ((2.*M_PI)*5.*i*ts - 1.0)     // f=5/N Hz
                + 0.6*cos ((2.*M_PI)*6.*i*ts - 2.0)     // f=6/N Hz
                + 0.7*cos ((2.*M_PI)*7.*i*ts - 3.0)     // f=7/N Hz
#endif
#if M > 4
                + 0.15*cos ((2.*M_PI)* 8.*i*ts + 2.5 )  // f=8/N Hz
                + 0.25*cos ((2.*M_PI)* 9.*i*ts + 1.5 )  // f=9/N Hz
                + 0.35*cos ((2.*M_PI)*10.*i*ts + 0.5 )  // f=10/N Hz
                + 0.45*cos ((2.*M_PI)*11.*i*ts + 0.25)  // f=11/N Hz
                + 0.55*cos ((2.*M_PI)*12.*i*ts - 0.25)  // f=12/N Hz
                + 0.65*cos ((2.*M_PI)*13.*i*ts - 0.5 )  // f=13/N Hz
                + 0.75*cos ((2.*M_PI)*14.*i*ts - 1.5 )  // f=14/N Hz
                + 0.85*cos ((2.*M_PI)*15.*i*ts - 2.5 )  // f=15/N Hz
#endif
                + 0.49;
        // Prevent from point N/2 being zero (here in the loop for readability
        // and to make the code at the end of this loop simpler)
        // NOTE that this will spoil the equality between magnitude/phase of
        // input harmonics and r/phi at test output.
        xr[N-1] = 0.13;
                                            //
#if M > 5
        {                                   // Create arbitrary test data for
            int  j;                         //   higher order FFTs
            for (j=(1<<4); j<N/2; ++j) {
                xr[i] += (double)rand()/((double)N*RAND_MAX)
                        * cos ((2.*M_PI)*j*i*ts + (double)rand()/RAND_MAX);
            }
        }
#endif

#ifndef NON_ZERO_IMAG_INPUT
        xi[i] = 0.;
#else
        xi[i] = i*2.0/N;
#endif
        xOri[i].r = xRef[i].r = xr[i];
        xOri[i].i = xRef[i].i = xi[i];
    }

    fftRef (xRef,N);

    //==========================================================================
    fputs (LOGO": Standard FFT Test\n", stderr);

#ifdef REAL_IN_OPTIMIZED
    for (i=0; i<N; ++i) {               // Create arbitrary input data, should
        xi[i] = rand();                 //   be ignored in the test object
    }
#endif

    fft (xr,xi);

#ifdef SYMM_OUT_OPTIMIZED
    for (i=N/2+1; i<N; ++i) {           // Reconstruct the omitted values
        xr[i] =  xr[N-i];               //   for the IFFT later
        xi[i] = -xi[N-i];
    }
#endif

    //--------------------------------------------------------------------------
    // Compare result
    for (i=0; i<N; ++i) {
        if (fabs(xr[i]-xRef[i].r) > EPS) {
            fprintf (stderr,
                     LOGO": at idx %d real: res: %13.6e <> ref: %13.6e\n",
                     i, xr[i], xRef[i].r);
            failed = 1;
        }
        if (fabs(xi[i]-xRef[i].i) > EPS) {
            fprintf (stderr,
                     LOGO": at idx %d imag: res: %13.6e <> ref: %13.6e\n",
                     i, xi[i], xRef[i].i);
            failed = 1;
        }
    }
#ifdef TEST_OUTPUT
    //--------------------------------------------------------------------------
    // Write test output to stdout
    puts ("Test Result                                  |Reference Result");
    for (i=0; i<N; ++i) {
        COMPLEX  x = {xr[i], xi[i]};
        double  r;
        double  phi;

        conv (x, &r, &phi);
        printf ("%4.1f: re%7.3f im%7.3f r%7.4f phi%7.4f",
                  i/(ts*N), xr[i], xi[i], r, phi);

        conv (xRef[i], &r, &phi);
        printf ("|%7.3f%7.3f %7.4f %7.4f\n", xRef[i].r, xRef[i].i, r, phi);
    }
    putchar ('\n');
#endif

    //==========================================================================
    fputs (LOGO": Inverse FFT Test\n", stderr);

#ifdef SYMM_IN_OPTIMIZED
    for (i=N/2+1; i<N; ++i) {           // Create arbitrary input data, should
        xr[i] = rand();                 //   be ignored in the test object
        xi[i] = rand();
    }
#endif

    ffti (xr,xi);

    for (i=0; i<N; ++i) {
        xr[i] *= 1./N;
        xi[i] *= 1./N;      // This can be omitted, if original sequence didn't
    }                       // contain imaginary values different from zero

    //--------------------------------------------------------------------------
    // Compare result
    for (i=0; i<N; ++i) {
        if (fabs(xr[i]-xOri[i].r) > EPS) {
            fprintf (stderr,
                     LOGO": at idx %d real: res: %13.6e <> ref: %13.6e\n",
                     i, xr[i], xOri[i].r);
            failed = 1;
        }
#ifndef REAL_OUT_OPTIMIZED
        if (fabs(xi[i]-xOri[i].i) > EPS) {
            fprintf (stderr,
                     LOGO": at idx %d imag: res: %13.6e <> ref: %13.6e\n",
                     i, xi[i], xOri[i].i);
            failed = 1;
        }
#endif
    }
#ifdef TEST_OUTPUT
    //--------------------------------------------------------------------------
    // Write test output to stdout
    puts ("Inverse FFT");
    puts ("Test Result                                    |Original Input Sequence");
    for (i=0; i<N; ++i) {
        COMPLEX  x = {xr[i], xi[i]};
        double  r;
        double  phi;

        conv (x, &r, &phi);
        printf ("%4.1f: re%8.4f im%7.4f r%7.4f phi%7.4f | reOri%8.4f imOri%7.4f\n",
                  i/(ts*N), xr[i], xi[i], r*(N/2), phi, xOri[i].r, xOri[i].i);
    }
#endif

    if (failed)  return 1;
    return 0;
}



//==============================================================================
// Reference FFT Algorithm
//

#define  SWAP(a,b)     do {COMPLEX t=(a);(a)=(b);(b)=t;} while (0)


void  fftRef (
    COMPLEX  x[],
    int      n
) {
    int     nm,mr,nn,m,k,istep,i,ii,jj;
    double  tr,ti,a,wr,wi;

    mr = 0;
    nn = n-1;

    for (m=1; m<=nn; ++m) {
        k = n;
        do {
            k /= 2;
        } while (mr+k > nn);
        mr = mr%k + k;
        if (mr > m)  SWAP(x[m],x[mr]);
    }
    for (k=1; k<n; k=istep) {
        istep = 2*k;
        for (m=0; m<k; ++m) {
            a  = M_PI*(-m)/k;
            wr = cos (a);
            wi = sin (a);
            ii = m;
            nm = (nn-m)/istep + m;
            for (i=m; i<=nm; ++i) {
                jj  = ii+k;
                tr  = wr*x[jj].r-wi*x[jj].i;
                ti  = wr*x[jj].i+wi*x[jj].r;
                x[jj].r = x[ii].r-tr;
                x[jj].i = x[ii].i-ti;
                x[ii].r += tr;
                x[ii].i += ti;
                ii += istep;
            }
        }
    }
}



//==============================================================================
// FFT Test Object
//

void  fft (
    FFT_TYPE  *xr,
    FFT_TYPE  *xi
) {
    FFT_TYPE  tr, ti;
#include "fft.c"
}



//==============================================================================
// IFFT Test Object
//

void  ffti (
    FFT_TYPE  *xr,
    FFT_TYPE  *xi
) {
    FFT_TYPE  tr, ti;
#include "ffti.c"
}



//==============================================================================
// Convert from cartesian coordinates to polar coordinates
//

void  conv (
    COMPLEX   x,
    double   *r,
    double   *phi
) {
    *r = sqrt (x.r*x.r+x.i*x.i)*(2./N);
    if (fabs(x.i) < fabs(x.r)) {
        if (x.r==0.) {
            *phi = 0.;
        } else {
            *phi = atan (x.i/x.r);
            if (x.r < 0.)  *phi += x.i>0 ? M_PI : -M_PI;
        }
    } else {
        if (x.i==0.) {
            *phi = 0.;
        } else {
            *phi = (x.i>=0. ? M_PI/2. : -M_PI/2.) - atan (x.r/x.i);
        }
    }
}


test/scripts/gcov2html.pl
#!/usr/bin/perl -w

my $version = 'Version 3';

################################################################################
#
# Source: gcov2html.pl
#
################################################################################
#
# Create documentation with
#  pod2html --infile=gcov2html.pl --outfile=gcov2html.pl.html --title "gcov2html.pl Manual"
#

################################################################################
# User Documentation

=head1 NAME

B<gcov2html.pl> - Convert the result of the GNU coverage testing tool to HTML
                  format


=head1 SYNOPSIS

B<gcov2html.pl> [I<Option>]... I<input-file>

Options:

[B<-v>] [B<-version>] [B<-h>] [B<--help>]


=head1 DESCRIPTION

gcov2html.pl converts the result of the GNU coverage testing tool (gcov) to
HTML format.  The result is written to STDOUT.


=head2 Arguments

=over 4

=item I<input-file>

Name of input file.

=back


=head2 Options

=over 4

=item B<-v>

=item B<-version>

Print version and exit..

=item B<-h>

=item B<--help>

Print usage information and terminate program.

=back


=head1 DIAGNOSTICS

Warnings and error messages are directed to STDERR.
There may be error messages from the Perl I/O system.


=head1 RETURN VALUE

Returns zero in case of normal successful completion.
In case of an error returns a value different from zero, normally the
value of errno (see PERL manual pages on "die").


=head1 BUGS

Not yet known.  If you encounter any error please report it to the author with
a brief description of the circumstances.


=head1 SEE ALSO

The documentation of the GNU coverage testing tool ("man 1 gcov" and
"man 1 gcc").


=head1 AUTHOR

jost.brachert@gmx.de


=head1 CHANGES

 Revision 4         Date: 15-Nov-2023           Author: jost
 '*' at end of flag for partly executed lines considered
 Legend added

 Revision 3         Date: 14-Feb-2018           Author: jost
 Documentation updated

 Revision 2         Date: 01-Oct-2017           Author: jost
 >, <, & will now be replaced by &gt;, &lt;, &amp;.

 Revision 1         Date: 14-Apr-2012           Author: jost
 Initial Revision


=head1 COPYING

gcov2html.pl is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation; either version 3 of the license, or (at your option) any later
version.

gcov2html.pl is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE. See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program, see file COPYING. If not, see L<https://www.gnu.org/licenses/>.

=cut

################################################################################


use strict;
use integer;

if (defined $ARGV[0]  and  substr($ARGV[0],0,2) eq '-v') {
    print __FILE__." $version\n";
    exit (0);
} elsif (not defined $ARGV[0]  or  substr($ARGV[0],0,1) eq '-') {
    print "Convert the result of gcov to html format\n".
          "usage: ".__FILE__." [options] <filename>\n".
          "            <filename>      Input filename.\n".
          "   options: -v  -version    Print version info and exit\n".
          "            -h  --help      Print usage information and exit\n";
    exit (0);
}

(my $title,) = $ARGV[0] =~ /^(.*)\.gcov$/;
print "<html><head><title>$title</title>\n";
print "<style type='text/css'><!--\n";
print "pre{font-size:x-small}\n";
print ".n{color:#4040FF}.e{color:#00D000}.p{color:#F08000}.u{color:#FF0000}.c{color:#808080}\n";
print "--></style>\n";
print "</head>\n";
print "<body><pre>\n";
while (my $inputline = <>) {
    chomp $inputline;
    my $match = (my $flag, my $lineno, my $line) = $inputline =~ /^([^:]+):\s*(\d+:+)(.*)$/;
    my $class = 'c';                            # Comment
    if    ($lineno =~ /^0:/ ) {$class = 'h';}   # Heading
    elsif ($flag =~ /#####/ ) {$class = 'u';}   # Unexecuted
    elsif ($flag =~ /\d$/   ) {$class = 'e';}   # Executed
    elsif ($flag =~ /\d\*?$/) {$class = 'p';}   # Partly executed

    $line =~ s/&/&amp;/g;
    $line =~ s/>/&gt;/g;
    $line =~ s/</&lt;/g;

    print "<span class='n'>$lineno</span> <span class='$class'>$line</span>\n";

    # Insert legend after "  -: 0:Runs:\d+"
    if ($inputline =~ /^\s+-:\s+0:Runs:\d+$/) {
        print "&nbsp;&nbsp;&nbsp;<span class='h'>Legend:</span>\n";
        print "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<span class='c'>No code</span>\n";
        print "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<span class='e'>Executed</span>\n";
        print "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<span class='p'>Partly executed</span>\n";
        print "&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<span class='u'>Unexecuted</span>\n";
    }
}
print "</pre></body></html>\n";


maketest.sh
#!/bin/bash
#
# Test script to test fftGen
#
# Version 1   2021 Jun 21st   jost.brachert@gmx.de
#
#-------------------------------------------------------------------------------
# Copyright (C) 2021  Jost Brachert, jost.brachert@gmx.de
#
# This program is free software: you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation; either version 3 of the license, or (at your option) any later
# version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with
# this program, see file COPYING. If not, see https://www.gnu.org/licenses/.
#

project="fftGen"

CFLAGS="-Wall"
CFLAGS="-lm"
source="../$project.c"

testscrdir="scripts"

sep="\n====\n"

cd test || exit

rm -v $project.gcda

#-------------------------------------------------------------------------------
# Create instrumented code
# => test/$project.gcno, test/$project
gcc -fprofile-arcs -ftest-coverage $CFLAGS $LDFLAGS -o $project $source

#-------------------------------------------------------------------------------
# Run tests
# => test/$project.gcda

date | tee stdout.log stderr.log

echo -e "${sep}Test help info output by short option"|\
    tee -a stderr.log >>stdout.log
./$project -h >>stdout.log 2>>stderr.log

echo -e "${sep}Test help info output by long option"|\
    tee -a stderr.log >>stdout.log
./$project --help >>stdout.log 2>>stderr.log

echo -e "${sep}Test version output by short option"|\
    tee -a stderr.log >>stdout.log
./$project -V >>stdout.log 2>>stderr.log

echo -e "${sep}Test version output by long option"|\
    tee -a stderr.log >>stdout.log
./$project --version >>stdout.log 2>>stderr.log

echo -e "${sep}Test no argument"|\
    tee -a stderr.log >>stdout.log
echo -e "    Expecting error message" >>stderr.log
./$project >>stdout.log 2>>stderr.log

echo -e "${sep}Test invalid option"|\
    tee -a stderr.log >>stdout.log
echo -e "    Expecting error message" >>stderr.log
./$project -y >>stdout.log 2>>stderr.log

echo -e "${sep}Test invalid argument"|\
    tee -a stderr.log >>stdout.log
echo -e "    Expecting error message" >>stderr.log
./$project y >>stdout.log 2>>stderr.log

echo -e "${sep}Test verbosity\nTest option assignment by =
Test option argument to -n is not power of 2"|\
    tee -a stderr.log >>stdout.log
echo -e "    Expecting error message" >>stderr.log
./$project -vn=7 >>stdout.log 2>>stderr.log

echo -e "${sep}Test invalid option argument"|\
    tee -a stderr.log >>stdout.log
echo -e "    Expecting error message" >>stderr.log
./$project -vn x >>stdout.log 2>>stderr.log

echo -e "${sep}Test 2-point FFT\nTest short option license output"|\
    tee -a stderr.log >>stdout.log
./$project -ln2 2>>stderr.log | tee fft.c >>stdout.log
./$project -in2 > ffti.c 2>>stderr.log
gcc $CFLAGS -DM=1 -DNON_ZERO_IMAG_INPUT -o fftTest fftTest.c 2>>stderr.log
if ! fftTest 2>>stderr.log ; then
    echo -e "\nTest failed, see stderr.log\n"
fi

echo -e "${sep}Test 4-point FFT\nTest long option license output
Test verbosity regarding license"|\
    tee -a stderr.log >>stdout.log
./$project -vln4 2>>stderr.log | tee fft.c >>stdout.log
./$project -in4 > ffti.c 2>>stderr.log
gcc $CFLAGS -DM=2 -DNON_ZERO_IMAG_INPUT -o fftTest fftTest.c 2>>stderr.log
if ! fftTest 2>>stderr.log ; then
    echo -e "\nTest failed, see stderr.log\n"
fi

echo -e "${sep}Test 8-point FFT\nTest no space before option argument"|\
    tee -a stderr.log >>stdout.log
./$project -n8  > fft.c  2>>stderr.log
./$project -in8 > ffti.c 2>>stderr.log
gcc $CFLAGS -DM=3 -DNON_ZERO_IMAG_INPUT -o fftTest fftTest.c 2>>stderr.log
if ! fftTest 2>>stderr.log ; then
    echo -e "\nTest failed, see stderr.log\n"
fi

echo -e "${sep}Test 16-point FFT\nTest space before option argument
Test -n and -i long option"|\
    tee -a stderr.log >>stdout.log
./$project -n 16  > fft.c  2>>stderr.log
./$project --inverse --points 16 > ffti.c 2>>stderr.log
gcc $CFLAGS -DM=4 -DNON_ZERO_IMAG_INPUT -o fftTest fftTest.c 2>>stderr.log
if ! fftTest 2>>stderr.log ; then
    echo -e "\nTest failed, see stderr.log\n"
fi

echo -e "${sep}Test 32-point FFT\nTest = before option argument
Test separate short options"|\
    tee -a stderr.log >>stdout.log
./$project -n=32 > fft.c  2>>stderr.log
./$project -i -n=32 > ffti.c 2>>stderr.log
gcc $CFLAGS -DM=5 -DNON_ZERO_IMAG_INPUT -o fftTest fftTest.c 2>>stderr.log
if ! fftTest 2>>stderr.log ; then
    echo -e "\nTest failed, see stderr.log\n"
fi

echo -e "${sep}Test 32-point FFT\nTest option -r"|\
    tee -a stderr.log >>stdout.log
./$project -rn32 > fft.c  2>>stderr.log
./$project -in32 > ffti.c 2>>stderr.log
gcc $CFLAGS -DM=5 -DREAL_IN_OPTIMIZED -o fftTest fftTest.c 2>>stderr.log
if ! fftTest 2>>stderr.log ; then
    echo -e "\nTest failed, see stderr.log\n"
fi

echo -e "${sep}Test 64-point FFT\nTest option -s"|\
    tee -a stderr.log >>stdout.log
./$project -sn64 > fft.c  2>>stderr.log
./$project -in64 > ffti.c 2>>stderr.log
gcc $CFLAGS -DM=6 -DSYMM_OUT_OPTIMIZED -o fftTest fftTest.c 2>>stderr.log
if ! fftTest 2>>stderr.log ; then
    echo -e "\nTest failed, see stderr.log\n"
fi

echo -e "${sep}Test 64-point FFT\nTest options -r and -s
Test -r and -s long option\nTest verbosity long option
Test verbosity regarding -r and -s"|\
    tee -a stderr.log >>stdout.log
./$project --verbose --real-in-opt --symm-out-opt -n64 > fft.c  2>>stderr.log
./$project -in64 > ffti.c 2>>stderr.log
gcc $CFLAGS -DM=6 -DREAL_IN_OPTIMIZED -DSYMM_OUT_OPTIMIZED -o fftTest fftTest.c 2>>stderr.log
if ! fftTest 2>>stderr.log ; then
    echo -e "\nTest failed, see stderr.log\n"
fi

echo -e "${sep}Test 128-point FFT\nTest option -m"|\
    tee -a stderr.log >>stdout.log
./$project -n128 > fft.c  2>>stderr.log
./$project -imn128 > ffti.c 2>>stderr.log
gcc $CFLAGS -DM=7 -DSYMM_IN_OPTIMIZED -o fftTest fftTest.c 2>>stderr.log
if ! fftTest 2>>stderr.log ; then
    echo -e "\nTest failed, see stderr.log\n"
fi

echo -e "${sep}Test 256-point FFT\nTest option -o"|\
    tee -a stderr.log >>stdout.log
./$project -n256 > fft.c  2>>stderr.log
./$project -ion256 > ffti.c 2>>stderr.log
gcc $CFLAGS -DM=8 -DREAL_OUT_OPTIMIZED -o fftTest fftTest.c 2>>stderr.log
if ! fftTest 2>>stderr.log ; then
    echo -e "\nTest failed, see stderr.log\n"
fi

echo -e "${sep}Test 512-point FFT\nTest options -m and -o
Test -m and -o long option\nTest verbosity regarding -m and -o"|\
    tee -a stderr.log >>stdout.log
./$project -n512 > fft.c  2>>stderr.log
./$project -vi --symm-in-opt --real-out-opt -n512 > ffti.c 2>>stderr.log
gcc $CFLAGS -DEPS=1.e-7 -DM=9\
 -DSYMM_IN_OPTIMIZED -DREAL_OUT_OPTIMIZED -o fftTest fftTest.c 2>>stderr.log
if ! fftTest 2>>stderr.log ; then
    echo -e "\nTest failed, see stderr.log\n"
fi

echo -e "${sep}Test 1024-point FFT\nTest options -r, -s, -m, -o\n"|\
    tee -a stderr.log >>stdout.log
./$project -rsn1024 > fft.c  2>>stderr.log
./$project -imon1024 > ffti.c 2>>stderr.log
gcc $CFLAGS -DEPS=1.e-7 -DM=10\
 -DREAL_IN_OPTIMIZED -DSYMM_OUT_OPTIMIZED -DSYMM_IN_OPTIMIZED -DREAL_OUT_OPTIMIZED\
 -o fftTest fftTest.c 2>>stderr.log
if ! fftTest 2>>stderr.log ; then
    echo -e "\nTest failed, see stderr.log\n"
fi

echo -e "${sep}Test usability for type float\n"|\
    tee -a stderr.log >>stdout.log
./$project -n32 > fft.c  2>>stderr.log
./$project -in32 > ffti.c 2>>stderr.log
gcc $CFLAGS -DFFT_TYPE=float -DEPS=1.e-5 -DM=5\
 -DNON_ZERO_IMAG_INPUT -o fftTest fftTest.c 2>>stderr.log
if ! fftTest 2>>stderr.log ; then
    echo -e "\nTest failed, see stderr.log\n"
fi

echo -e "test script finished\n"

#-------------------------------------------------------------------------------
# Create coverage report in ASCII format
# => test/fftGen.c.gcov
gcov -o $project.gcno ../$project.c

# Create coverage report in HTML format
# => test/fftGen.cov.html
$testscrdir/gcov2html.pl $project.c.gcov > $project.cov.html

# Verify the test results
if ! diff -b -I "\([0-9][0-9]:\)\{2\}[0-9][0-9].*[0-9]\{4\}" stdout.log refs/stdout.log ; then
    echo -e "test failed: stdout.log differs from reference"
    exit 1
fi
if ! diff -b -I "\([0-9][0-9]:\)\{2\}[0-9][0-9].*[0-9]\{4\}" stderr.log refs/stderr.log ; then
    echo -e "test failed: stderr.log differs from reference"
    exit 1
fi

echo -e "test passed"
